crates/data-generator/src/blocks/output.rs

```diff
@@ -0,0 +1,586 @@
+                }
+            });
+        }
+    }
+}
+
+mod block_exports {
+    use proc_macro2::TokenStream;
+    use quote::{quote, ToTokens};
+
+    use crate::name::Name;
+
+    pub struct BlockExports<'a> {
+        mod_block_names: Vec<&'a Name>,
+    }
+
+    impl<'a> BlockExports<'a> {
+        pub fn new(blocks: impl IntoIterator<Item = &'a Name>) -> Self {
+            Self {
+                mod_block_names: blocks.into_iter().collect(),
+            }
+        }
+    }
+
+    impl ToTokens for BlockExports<'_> {
+        fn to_tokens(&self, tokens: &mut TokenStream) {
+            let mut exports = TokenStream::new();
+            for name in self.mod_block_names.iter() {
+                let block_mod_name = name.snake_case();
+                let block_struct_name = name.pascal_case();
+                exports.extend(quote!(#block_mod_name::Block as #block_struct_name,))
+            }
+
+            // "pub use self::cave_vines::Block as CaveVines;"
+            tokens.extend(quote!(pub use self::{#exports};))
+        }
+    }
+}
+
+mod block_mod {
+    use proc_macro2::TokenStream;
+    use quote::{quote, ToTokens};
+
+    use crate::{blocks::model::Block, name::Name};
+
+    pub struct BlockMod<'a> {
+        pub name: &'a Name,
+        pub block: &'a Block,
+    }
+
+    impl ToTokens for BlockMod<'_> {
+        fn to_tokens(&self, tokens: &mut TokenStream) {
+            let block_mod_name = self.name.snake_case();
+            let block_code = self.gen_block_code();
+            let props_code = self.gen_props_code();
+
+            tokens.extend(quote! {
+                pub mod #block_mod_name {
+                    #block_code
+                    #props_code
+                }
+            });
+        }
+    }
+
+    impl BlockMod<'_> {
+        fn gen_block_code(&self) -> TokenStream {
+            let mut tokens = TokenStream::new();
+            super::block_struct::BlockStruct::new(self.block).to_tokens(&mut tokens);
+            super::block_convert::BlockConvert::from(self.block).to_tokens(&mut tokens);
+            tokens
+        }
+
+        fn gen_props_code(&self) -> TokenStream {
+            use super::{prop_convert::*, prop_default::*, prop_enum::*};
+
+            let mut tokens = TokenStream::new();
+            for (prop_name, variants) in self.block.properties.iter() {
+                let (_, def_state) = self.block.states.default();
+                let def_value = def_state
+                    .properties
+                    .get(prop_name)
+                    .expect("property is present");
+
+                PropertyEnum::new(prop_name, variants).to_tokens(&mut tokens);
+                PropertyDefault::new(prop_name, def_value).to_tokens(&mut tokens);
+                PropertyConvert::new(prop_name, variants).to_tokens(&mut tokens);
+            }
+            tokens
+        }
+    }
+}
+
+mod block_struct {
+    use proc_macro2::{Delimiter, Group, Ident, Punct, Spacing, TokenStream};
```

Comment:
```suggestion
    use proc_macro2::*;
```
Waste of space

Reply:
Do we really want to introduce asterisk wildcards in Rjacraft? I don't mind wasting extra 200 bytes of disk space. @JarvisCraft